src/cloudant/search.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+# Copyright (c) 2016 IBM. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+API module for composing and executing Cloudant Search indexes.
+"""
+from ._common_util import codify
+from cloudant.error import CloudantException
+
+
+class SearchIndex(dict):
+    """
+    Encapsulates a search index as a dictionary based object, exposing the
+    search index function as an attribute and supporting query/data access
+    through the search endpoint.  A Search object is instantiated with a
+    reference to a DesignDocument and is typically used as part of the
+    :class:`~cloudant.design_document.DesignDocument`
+    search index management API.
+    """
+    def __init__(
+            self,
+            ddoc,
+            search_index_name,
+            search_index=None,
+            **kwargs
+    ):
+        super(SearchIndex, self).__init__()
+        self._design_doc = ddoc
+        self._search_index_name = search_index_name
+        self['index'] = codify(search_index)
+        if kwargs:
+            super(SearchIndex, self).update(kwargs)
+
+    @property
+    def url(self):
+        """
+        Constructs and returns the Cloudant Search URL.
+
+        :returns: Search URL
+        """
+        if not self._search_index_name:
+            return '/'.join([
+                self._design_doc.document_url,
+                '_search',
+                self._search_index_name
+            ])
+        else:
+            msg = 'Search index name is None or was not provided.'
+            raise CloudantException(msg)
+
```

Comment:
There a few things that need to be mentioned here:

1.  I think your `if/else` logic is reversed.  Which makes me wonder about your tests if this is not failing.
2.  There does not seem to be a point for the `if/else` block since the `search_index_name` is required in the constructor and will fail there.  The only purpose this check serves is if someone specifically entered in a series of blank spaces as a `search_index_name` but we don't do this type of check anywhere else in the library and I think if at some point you want to add this level of validation then it might make sense to make that decision for the entire library.
3.  Finally, I don't see a need for the `url` property here at all unless you plan to make a `SearchIndex` callable.  I don't think that is what we agreed to before but things may have changed since we spoke last.  In either case, this url property doesn't seem to add any value from a search index management stand point.  So I think it makes sense to remove it and include it where appropriate when you implement the search query execution piece of this work in a later PR.